updater/sis-updater-worker/src/updater/updateStudents/attainments.js

```diff
@@ -135,53 +135,44 @@ const updateAttainments = async (
-      const courseUnits = courses.filter(c => c.code === attIdToCourseCode[att.id])
-
-      const isAfterStartAndBeforeEnd = (startDate, endDate, date) => {
-        const dateToCompare = new Date(date)
-        return new Date(startDate) <= dateToCompare && (!endDate || dateToCompare < new Date(endDate))
-      }
-
-      let courseUnit = courseUnits.find(({ validity_period }) =>
-        isAfterStartAndBeforeEnd(validity_period.startDate, validity_period.endDate, att.attainment_date)
-      )
-
-      // Sometimes registrations are fakd, see attainment hy-opinto-141561630.
-      // The attainmentdate is outside of all courses, yet should be mapped.
-      // Try to catch suitable courseUnit for this purpose
-      if (!courseUnit) {
-        courseUnit = courseUnits.find(({ validity_period }) =>
-          isAfterStartAndBeforeEnd(validity_period.startDate, validity_period.endDate, att.registration_date)
-        )
-      }
+      let courseUnit = courses.find(c => c.id === att.course_unit_id)
 
       // If there's no suitable courseunit, there isn't courseunit available at all.
       // --> Course should be created, if it doesn't exist in sis db
       if (!courseUnit) {
         const parsedCourseCode = attIdToCourseCode[att.id]
-        // see if course exists
-        const course = await Course.findOne({ where: { code: parsedCourseCode } })
-
-        // If course doesn't exist, create it
-        if (!course) {
-          coursesToBeCreated.set(parsedCourseCode, {
-            id: parsedCourseCode,
-            name: att.name,
-            code: parsedCourseCode,
-            coursetypecode: att.study_level_urn,
-            course_unit_type: att.course_unit_type_urn,
-          })
+        courseUnit = {
+          id: att.id,
+          name: att.name,
+          code: att.code,
+          latest_instance_date: att.attainment_date,
+          is_study_module: !(
+            att.type in
+            [
+              'StudyModuleAttainment',
+              'CustomStudyModuleAttainment',
+              'DegreeProgrammeAttainment',
+              'CustomDegreeProgrammeAttainment',
+            ]
+          ),
+          coursetypecode: att.study_level_urn,
+          max_attainment_date: att.attainment_date,
+          min_attainment_date: att.attainment_date,
+          substitutions: [],
+          course_unit_type: att.course_unit_type_urn,
         }
+        coursesToBeCreated.set(att.id, courseUnit)
 
         // see if course has provider
         const courseProvider = await CourseProvider.findOne({
           where: {
-            coursecode: course ? course.id : parsedCourseCode,
+            coursecode: att.course_unit_id,
           },
         })
 
         // If there's no courseprovider, try to create course provider
         if (!courseProvider) {
+          // parsedCourseCode but method wants groupId ? sis-importer model has no groupId :<
           const mapCourseProvider = courseProviderMapper(parsedCourseCode)
```

Comment:
There's now a little problem regarding how new course providers are created. `coursecode` in the course_providers table should reference the `id` column in the course table but now in custom attainments the id is actually the id from the attainment.

So maybe we should replace `parsedCourseCode` with `att.id`? This seems to fix the problem at least for us.

Reply:
Changed and seems to be working still